src/kernels/raw/finder/localKernelFinder.node.ts

```diff
@@ -4,134 +4,226 @@
-                .then((items) => {
-                    kernelsFromCache = items;
-                    updateCache = false;
-                })
-                .catch(noop);
+        }
+        await this.writeToCache(kernels);
+        this._initializeResolve();
+    }
+
+    private async updateCache(resource: Resource) {
+        this._cacheUpdateCancelTokenSource?.dispose();
+        const updateCacheCancellationToken = new CancellationTokenSource();
+        this._cacheUpdateCancelTokenSource = updateCacheCancellationToken;
+
+        if (resource) {
+            let kernels: LocalKernelConnectionMetadata[];
 
             try {
-                const kernelsWithoutCacheDeferred = createDeferredFromPromise(kernelsWithoutCachePromise);
-                try {
-                    await Promise.race([kernelsFromCachePromise, kernelsWithoutCacheDeferred.promise]);
-                } catch (ex) {
-                    // If we failed to get without cache, then await on the cache promise as a fallback.
-                    if (kernelsWithoutCacheDeferred.rejected) {
-                        await kernelsFromCachePromise;
-                    } else {
-                        throw ex;
-                    }
-                }
-                // If we finish the cache first, and we don't have any items, in the cache, then load without cache.
-                if (Array.isArray(kernelsFromCache) && kernelsFromCache.length > 0) {
-                    kernels = kernelsFromCache;
-                } else {
-                    kernels = await kernelsWithoutCachePromise;
-                    updateCache = true;
-                }
+                kernels = await this.listKernels(resource, updateCacheCancellationToken.token);
             } catch (ex) {
                 traceError(`Exception loading kernels: ${ex}`);
+                return;
             }
+
+            // update resource cache
+            const resourceCacheKey = this.getResourceCacheKey(resource);
+            this.resourceCache.set(resourceCacheKey, kernels);
+
+            this._onDidChangeKernels.fire();
+        } else if (workspace.workspaceFolders) {
+            await Promise.all(
+                workspace.workspaceFolders.map((folder) => () => {
+                    return new Promise<void>(async (resolve) => {
+                        try {
+                            let kernels = await this.listKernels(folder.uri, updateCacheCancellationToken.token);
+                            await this.writeToCache(kernels);
+                            const resourceCacheKey = this.getResourceCacheKey(folder.uri);
+                            this.resourceCache.set(resourceCacheKey, kernels);
+                        } catch (ex) {
+                            traceError(`Exception loading kernels: ${ex}`);
+                        }
+                        resolve();
+                    });
+                })
+            );
+        }
+
+        this._onDidChangeKernels.fire();
```

Comment:
Same as earlier, check if the cancel token has been cancelled and then return

Reply:
Basically we shouldn't be updating the list with an empty list and triggering the event, else the callers could end up getting an empty list
all because we have this method getting executed twice.